apps/emqx_oracle/src/emqx_oracle.erl

```diff
@@ -0,0 +1,375 @@
+                {ok, NState} ->
+                    %% return new state with prepared statements
+                    {connected, NState}
+            end;
+        false ->
+            disconnected
+    end.
+
+do_get_status(Conn) ->
+    ok == element(1, jamdb_oracle:sql_query(Conn, "select 1 from dual")).
+
+do_check_prepares(#{prepare_sql := Prepares}) when is_map(Prepares) ->
+    ok;
+do_check_prepares(State = #{pool_name := PoolName, prepare_sql := {error, Prepares}}) ->
+    {ok, Sts} = prepare_sql(Prepares, PoolName),
+    {ok, State#{prepare_sql => Prepares, prepare_statement := Sts}}.
+
+%% ===================================================================
+
+oracle_host_options() ->
+    ?ORACLE_HOST_OPTIONS.
+
+connect(Opts) ->
+    Password = emqx_secret:unwrap(proplists:get_value(password, Opts)),
+    NewOpts = lists:keyreplace(password, 1, Opts, {password, Password}),
+    jamdb_oracle:start_link(NewOpts).
+
+sql_query_to_str(SqlQuery) ->
+    emqx_plugin_libs_rule:str(SqlQuery).
+
+sql_params_to_str(Params) when is_list(Params) ->
+    lists:map(
+        fun
+            (false) -> "0";
+            (true) -> "1";
+            (Value) -> emqx_plugin_libs_rule:str(Value)
+        end,
+        Params
+    ).
+
+query(Conn, SQL, Params) ->
+    Ret = jamdb_oracle:sql_query(Conn, {sql_query_to_str(SQL), sql_params_to_str(Params)}),
+    ?tp(oracle_query, #{conn => Conn, sql => SQL, params => Params, result => Ret}),
+    handle_result(Ret).
+
+execute_batch(Conn, SQL, ParamsList) ->
+    ParamsListStr = lists:map(fun sql_params_to_str/1, ParamsList),
+    Ret = jamdb_oracle:sql_query(Conn, {batch, sql_query_to_str(SQL), ParamsListStr}),
+    ?tp(oracle_batch_query, #{conn => Conn, sql => SQL, params => ParamsList, result => Ret}),
+    handle_result(Ret).
+
+parse_prepare_sql(Config) ->
+    SQL =
+        case maps:get(prepare_statement, Config, undefined) of
+            undefined ->
+                case maps:get(sql, Config, undefined) of
+                    undefined -> #{};
+                    Template -> #{<<"send_message">> => Template}
+                end;
+            Any ->
+                Any
+        end,
+    parse_prepare_sql(maps:to_list(SQL), #{}, #{}).
+
+parse_prepare_sql([{Key, H} | T], Prepares, Tokens) ->
+    {PrepareSQL, ParamsTokens} = emqx_plugin_libs_rule:preproc_sql(H, ':n'),
+    parse_prepare_sql(
+        T, Prepares#{Key => PrepareSQL}, Tokens#{Key => ParamsTokens}
+    );
+parse_prepare_sql([], Prepares, Tokens) ->
+    #{
+        prepare_sql => Prepares,
+        params_tokens => Tokens
+    }.
+
+init_prepare(State = #{prepare_sql := Prepares, pool_name := PoolName}) ->
+    {ok, Sts} = prepare_sql(Prepares, PoolName),
+    State#{prepare_statement := Sts}.
+
+prepare_sql(Prepares, PoolName) when is_map(Prepares) ->
+    prepare_sql(maps:to_list(Prepares), PoolName);
+prepare_sql(Prepares, PoolName) ->
+    Data = do_prepare_sql(Prepares, PoolName),
+    {ok, _Sts} = Data,
```

Comment:
Nit: seems like this will never fail, and `LastSts` in `do_prepare_sql` below is not used anywhere?

Reply:
Ignore me.